ducktape/mark/_mark.py

```diff
@@ -0,0 +1,359 @@
+
+    return ignorer
+
+
+def _inject(*args, **kwargs):
+    """Inject variables into the arguments of a function or method.
+    This is almost identical to decorating with functools.partial, except we also propagate the wrapped
+    function's __name__.
+    """
+
+    def injector(f):
+        assert callable(f)
+
+        @functools.wraps(f)
+        def wrapper(*w_args, **w_kwargs):
+            return functools.partial(f, *args, **kwargs)(*w_args, **w_kwargs)
+
+        wrapper.args = args
+        wrapper.kwargs = kwargs
+        wrapper.function = f
+
+        return wrapper
+    return injector
+
+
+class MarkedFunctionExpander(object):
+    """This class helps expand decorated/marked functions into a list of test context objects. """
+    def __init__(self, session_context=None, module=None, cls=None, function=None):
+        self.session_context = session_context
+        self.module = module
+        self.cls = cls
+        self.function = function
+
+    def expand(self, test_parameters=None):
+        """Inspect self.function for marks, and expand into a list of test context objects useable by the test runner.
+        """
+        f = self.function
+        if not hasattr(f, "marks"):
+            return [TestContext(self.session_context, self.module, self.cls, self.function)]
+
+        expanded = []
+        parametrize_marks = [m for m in f.marks if m.name == PARAMETRIZED.name]
+        matrix_marks = [m for m in f.marks if m.name == MATRIX.name]
+        ignore_marks = [m for m in f.marks if m.name == IGNORE.name]
+
+        ignore_list = [m.injected_args for m in ignore_marks]
+        ignore_all = None in ignore_list
+
+        if test_parameters is not None:
+            # User has specified that they want to run tests with specific parameters
+            # Return a list consisting of a single test context object with the specified test_parameters whether
+            # or not the function has parametrize decorators
+            injected_fun = _inject(**test_parameters)(self.function)
+            context = TestContext(self.session_context, self.module, self.cls, injected_fun, test_parameters)
+            context.ignore = None in ignore_list or test_parameters in ignore_list
+            return [context]
+
+        # unparametrized function is expanded into a list consisting of a single test context object
+        if len(parametrize_marks) == 0 and len(matrix_marks) == 0:
+            context = TestContext(self.session_context, self.module, self.cls, self.function)
+            context.ignore = ignore_all
+            return [context]
+
+        # expand parametrize marks, adding "ignore" metadata where necessary
+        for m in parametrize_marks:
+            injected_fun = _inject(**m.injected_args)(self.function)
+
+            context = TestContext(self.session_context, self.module, self.cls, injected_fun, m.injected_args)
+            context.ignore = None in ignore_list or m.injected_args in ignore_list
```

Comment:
I think these pieces overriding the `ignore` flag are a great example of how this is going to get really messy if we don't set these annotations up to compose well from the get go.

I'd suggest changing all of these to make use of a `TestContext` constructor that accepts `**kwargs` to be able to copy the entire original dict, with updates applied to keys that you know need to be overridden (the function and args in this case). This ensures all existing properties will be propagated, which makes a lot of cases work out naturally when composed (e.g., even if you put `@ignore()` first, the flag will be propagated to all `TestContext`s even if they are further expanded; if you included arguments, the flag + arguments can be copied to all `TestContext`s such that the ignore flag is still only applied to `TestContext`s with matching keyword args).